### DISCUSSION TITLE
Project for potential conversion to LoRa & FDRS

### DISCUSSION THREAD

**Discussion by Astrobirder** [timmbogner/Farm-Data-Relay-System#181] [Q&A]
Hello,

A friend and I are about to expand an existing "farm" monitoring system from 4 arduino nodes using HC-12 & a Raspberry Pi to a system that includes all of that and some nodes 2km away (not perfect line of sight--there are woodlands between, though the ground is fairly flat).   Today's system is all local to one site and all nodes have access to "mains" A/C power / DC conversion.

We are considering FDRS and LoRa radios, but neither of us have used LoRa before (though we are both retired electrical engineers with amateur radio licenses).

Pictures can convey more, so here is today: 

![Slide1](https://github.com/timmbogner/Farm-Data-Relay-System/assets/13646631/0b04e22f-703f-456a-9e57-e616e50c0fe4)

And here are 2 guesses at what it might look like if we move to FDRS, though not sure we have the right H/W selected.

![Slide2](https://github.com/timmbogner/Farm-Data-Relay-System/assets/13646631/ff8371aa-581a-4ef6-8c22-5e2f939bb613)

![Slide3](https://github.com/timmbogner/Farm-Data-Relay-System/assets/13646631/e7d64f4d-faa2-45c8-bc70-db8adb6bdaf3)

Questions:

1. Is it possible to put EVERYTHING onto the Pi, and not bother with an ESP gateway?  We aren't using ESPNow in this system, as we're building it ourselves.

2. We are considering the Adafruit RFM95 Breakout boards to begin experiments.  It appears that these are supported by the RadioLib Arduino library, but I don't know if that can run on an Arduino Mega.  We're using the Mega currently, as that's what we have AND we need a fair bit of I/O for one or two of the nodes (6 A/D for temperatures, 1 A/D for ambient light sensor, 6 discrete outputs, and 6 outs + 6 Ins for HX711 scale interfaces).   We haven't found an ESP device that can handle that.   Any better ideas?

3. Note that the new REMOTE node is 2 km away and needs to be SOLAR powered, so we'll need to do some charge control and panel interfacing.  Initially it isn't as large as the node detailed above, but it might grow to that.  We are currently thinking if we can keep the Arduino/ESP architecture the same between the remote and node above, it would simplify code maintenance.  Any ideas on a better selection for the remote microcontroller?

4. We are considering building yagi antennas for both ends of the distant connection, but are concerned that many of the Adafruit RFM95 pages say the range limit is 2km WITH UNIDIRECTIONAL antennas.  We don't know if that's what Adafruit means, or if it is something else.  RFM95s appear to be limited to 100mW out, even though the FCC (This system if for USA) limit is 1 W, with 4W ERP.  Should we be considering something else, or will the yagi antennas be enough?

Thank you in advance for your comments.

**Comment by timmbogner** [comment]
Sounds cool!

I think with 2km and some woodlands you may have to play with some LoRa radio's configurations, especially spreading factor, to make the distance. I'd suggest trying out[ this tutorial ](https://hackaday.io/project/190006-range-testing-with-fdrs-and-node-red)I made to plot a map of your range with different configurations
Since you mention you have radio licenses, that may open you up to using the 433mhz band hardware if you're in areas with legal restrictions.

1. Currently, there is no ready-made method for using a Raspberry Pi with a LoRa module directly connected. All of the code runs on microcontrollers. Option 2 is more on track. In my eyes, the best setup would be a LoRa module connected to an ESP or other microcontroller, which is connected directly to the RPi via UART.
If you needed to separate the LoRa receiver and the RPi, I would use MQTT using either an ethernet-enabled ESP32 or regular WiFi. I don't see a need for ESP-NOW from your diagram.
2. I think that RadioLib should work alright on Mega, but there is not enough memory to run FDRS in its current state. There may be some easy optimization, but it's beyond the scope of things I reasonably can or will accomplish in the next couple months. The main culprit is that I think my use of the Arduino JSON parser is a bit inefficient.
That is certainly a lot of IO as well. My first thought is that maybe you can either multiplex those temp sensors, or use DS18B20 instead. You could also split that device into two ESP32s. Maybe split those sensors up and send them separately using ESP-NOW instead of LoRa, since 10m is pretty short. There would be an extra step involved if you're using WiFi to connect to the RPi, but if you're using a UART or ethernet link, this would be a very easy way to go.
RP2040 is also supported by FDRS and RadioLib, would that have the IOs needed?
3. I just started experimenting with solar panels this year, so take my observations with a grain of salt. I think there are probably other architectures that have lower sleep current consumption, but I mostly know about ESP32 and 8266. I haven't had much trouble with powering sensor nodes from solar. If they spend most of their time asleep, they can usually make it through the night on even a small battery. Making sure all of your sensors don't drawn current when sleeping is a biggie. It really comes down to the question of how much current the sensors will draw, how long it will take to complete your measurements, and how often you take them. I think the ESP32 S3 is touted to have the best current consumption in sleep, but I'm a bit rusty.
4. You probably know more about the antenna side of things than myself. Most people use smaller-sized antennas with decent success, so I would imagine that a nicer one would only improve things! I think that 2km hop through trees might be a struggle though. Again, choosing a 433mhz hardware setup could also solve some range issues.

I'm a bit rusty on a lot of this, as I haven't had a lot of time to spend on FDRS lately. I plan to return to work in a month or so to fix the outstanding issues, then work on new features. I'm always around to answer questions, though.
Thanks for checking out FDRS!

**Comment by bvwelch** [comment]
Hi guys! I have done some similar tinkering here in Alabama but I have no sure-fire answers, sorry. I also want to use this FDRS but I was already in the midst of using 2 of the adafruit feather m0 lora boards, to keep track of my postal mailbox which is about 1000 ft away.   I also bought one of the adafruit rfm95 breakout boards and connected it to a esp32 and it seems to work OK.   

I have used some of the old 328p avr/arduino boards but I am not familiar with your Mega.

However, I have used raspberry pi for some years now.  And I recently read about, but have not yet tested, that you can use adafruit's 'circuitpython' libraries from your Pi, which should in theory work with the adafruit lora breakout.   

This fdrs is a great source of lora info.  another one is 'stuartsprojects' -- lots of good info and code and schematics.    And just last nite, I watched an old digikey youtube series where the guy did a low power 328p / lora project, with the other end being the raspberry pi and a adafruit lora breakout.  and there is always the 'swiss guy' who has lots of great info.  he is also a ham.    73 de ab4yd.  ps  I have been pondering 433 lora too but for now just using 915.

https://stuartsprojects.github.io/
https://www.digikey.com/en/maker/projects/arduino-low-power-creating-a-lora-iot-node-that-runs-off-batteries/6936a0de7db249fe90a5edc661b618d3
https://www.youtube.com/@AndreasSpiess
https://bvwelch.github.io/lora-radio/
